org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semanticHighlight/SemanticHighlightingRegistry.java

```diff
@@ -0,0 +1,276 @@
+import org.eclipse.lsp4j.VersionedTextDocumentIdentifier;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.util.SemanticHighlightingTokens;
+import org.eclipse.lsp4j.util.SemanticHighlightingTokens.Token;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.ISemanticHighlightingCalculator;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.MergingHighlightedPositionAcceptor;
+import org.eclipse.xtext.ide.server.Document;
+import org.eclipse.xtext.ide.server.ILanguageServerAccess.Context;
+import org.eclipse.xtext.ide.server.UriExtensions;
+import org.eclipse.xtext.resource.IResourceServiceProvider;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.xbase.lib.Extension;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xbase.lib.ListExtensions;
+import org.eclipse.xtext.xbase.lib.util.ToStringBuilder;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+
+/**
+ * Shared semantic highlighting manager per language server. Responsible for converting the semantic highlighted ranges
+ * into the LSP standard by producing a compact, {@code base64} encoded token string.
+ */
+public class SemanticHighlightingRegistry {
+	/**
+	 * Reserved TextMate scope identifier for styles that cannot be handled.
+	 */
+	public static final String UNKNOWN_SCOPE = "unknown.xtext";
+
+	/**
+	 * TextMate scopes indicating an unhandled style ID to TextMate scope mapping.
+	 */
+	public static final List<String> UNKNOWN_SCOPES = Collections
+			.singletonList(SemanticHighlightingRegistry.UNKNOWN_SCOPE);
+
+	@Inject
+	@Extension
+	private UriExtensions uriExtensions;
+
+	/**
+	 * Lookup table for all known TextMate scopes.
+	 */
+	protected BiMap<Integer, List<String>> scopes;
+
+	protected LanguageClient client;
+
+	public void initialize(Iterable<? extends IResourceServiceProvider> allLanguages, ClientCapabilities capabilities,
+			LanguageClient client) {
+		Preconditions.checkState(this.client == null, "Already initialized.");
+		TextDocumentClientCapabilities textDocument = capabilities == null ? null : capabilities.getTextDocument();
+		SemanticHighlightingCapabilities semanticHighlightingCapabilities = textDocument == null ? null
+				: textDocument.getSemanticHighlightingCapabilities();
+		boolean semanticHighlighting = semanticHighlightingCapabilities == null ? false
+				: semanticHighlightingCapabilities.getSemanticHighlighting();
+		ImmutableBiMap.Builder<Integer, List<String>> builder = ImmutableBiMap.builder();
+		if (semanticHighlighting) {
+			Set<List<String>> set = IterableExtensions.toSet(IterableExtensions.filter(
```

Comment:
Please either go for Stream API, FluentIterable or traditional for loops. But these inversely nested static methods make it really hard to understand the code.

Reply:
Yes, it is hard to understand. Will try to refactor it.

Reply:
Refactored to traditional for loops. Please verify that my implementation maps well to the old implementation.

Reply:
Seems our tests are not THAT bad ... seems they catched a bug in my implementation ... will check.

Reply:
Have rewritten it using Streams for a more direct mapping to the old code and to fix the tests. Please review again.